README.md
# hydra_login2f

*hydra_login2f* is a secure login provider for [ORY Hydra OAuth2
Server](https://github.com/ory/hydra). *hydra_login2f* implements
two-factor authentication via email.


## Installation

*hydra_login2f* can be deployed directly from a [docker
image](https://hub.docker.com/r/epandurski/hydra_login2f/). You can
find a working example in the `example/` directory.


## Configuration

*hydra_login2f*'s behavior can be tuned with environment
variables. Here are the most important settings with their default
values:

``` shell
# The port on which `hydra_login2f` will run.
PORT=8000

# The path to the login page (ORY Hydra's `OAUTH2_LOGIN_URL`):
LOGIN_PATH='/login'

# The path to the dummy consent page (ORY Hydra's `OAUTH2_CONSENT_URL`).
# `hydra_login2f` implements a dummy consent page, which accepts all
# consent requests unconditionally, without showing any UI to the user.
# This is sometimes useful, especially during testing.
CONSENT_PATH='/consent'

# The prefix added the user ID to form the Oauth2 subject field. For
# example, if SUBJECT_PREFIX='user:', the OAuth2 subject for the user
# with ID=1234 would be 'user:1234'.
SUBJECT_PREFIX=''

# Set this to a random, long string, and keep it secret.
SECRET_KEY='dummy-secret'

# Set this to the name of your site, as it is known to your users.
SITE_TITLE='My site name'

# Set this to an URL that tells more about your site.
ABOUT_URL='https://github.com/epandurski/hydra_login2f'

# Optional URL for a custom CSS style-sheet:
STYLE_URL=''

# Whether to issue recovery codes to your users for additional security
# ('True' or 'False'). It is probably a good idea to use recovery codes
# if the account to your service might be more important to your users
# than their email account.
USE_RECOVERY_CODE=True

# Set this to the URL for ORY Hydra's admin API.
HYDRA_ADMIN_URL='http://hydra:4445'

# Set this to the URL for your Redis server instance. It is highly
# recommended that your Redis instance is backed by disk storage. If not so,
# your users might be inconvenienced when your Redis instace is restarted.
REDIS_URL='redis://localhost:6379/0'

# Set this to the URL for your SQL database server instance. PostgreSQL
# and MySQL are supported out of the box. Example URLs:
# - postgresql://user:pass@servername/dbname
# - mysql+mysqlconnector://user:pass@servername/dbname
SQLALCHEMY_DATABASE_URI=''

# The size of the database connection pool. If not set, defaults to the
# engine’s default (usually 5).
SQLALCHEMY_POOL_SIZE=None

# Controls the number of connections that can be created after the pool
# reached its maximum size (`SQLALCHEMY_POOL_SIZE`). When those additional
# connections are returned to the pool, they are disconnected and discarded.
SQLALCHEMY_MAX_OVERFLOW=None

# Specifies the connection timeout in seconds for the pool.
SQLALCHEMY_POOL_TIMEOUT=None

# The number of seconds after which a connection is automatically recycled.
# This is required for MySQL, which removes connections after 8 hours idle
# by default. It will be automatically set to 2 hours if MySQL is used.
# Some backends may use a different default timeout value (MariaDB, for
# example).
SQLALCHEMY_POOL_RECYCLE=None

# SMTP server connection parameters. You should set `MAIL_DEFAULT_SENDER`
# to the email address from which you send your outgoing emails to users,
# "My Site Name <no-reply@my-site.com>" for example.
MAIL_SERVER='localhost'
MAIL_PORT=25
MAIL_USE_TLS=False
MAIL_USE_SSL=False
MAIL_USERNAME=None
MAIL_PASSWORD=None
MAIL_DEFAULT_SENDER=None

# Parameters for Google reCAPTCHA 2. You should obtain your own public/private
# key pair from www.google.com/recaptcha, and put it here.
RECAPTCHA_PUBLIC_KEY='6Lc902MUAAAAAJL22lcbpY3fvg3j4LSERDDQYe37'
RECAPTCHA_PIVATE_KEY='6Lc902MUAAAAAN--r4vUr8Vr7MU1PF16D9k2Ds9Q'

# Set this to the number of worker processes for handling requests -- a
# positive integer generally in the 2-4 * $NUM_CORES range.
GUNICORN_WORKERS=2

# Set this to the number of worker threads for handling requests. (Runs
# each worker with the specified number of threads.)
GUNICORN_THREADS=1
```


CHANGES.md
Changelog
=========


Version 0.9.5
-----------

- Fixed a potential security issue caused by storing potentially
  sensitive data in Redis. Now only cryptographic hashes are stored
  instead of the data.

- Improved README


Version 0.9.4
-----------

- Added database connection pooling configuration settings
- Improved default logging configuration
- Improved Dockerfile


Version 0.9.3
-----------

- Do not require verification code after password change
- Use python:3.7-alpine for the docker image


Version 0.9.2
-----------

- Added `SUBJECT_PREFIX` parameter
- Minor improvement in the `example/`


Version 0.9.1
-----------

- Fixed bug in MySQL support



Version 0.9
-----------

- Initial public release


hydra_login2f/config.py
from flask_env import MetaFlaskEnv
from os import environ

SUPPORTED_LANGUAGES = {'en': 'English', 'bg': 'Български'}


def _get_language_choices(fallback):
    languages = environ.get('LANGUAGES', fallback)
    languages = [l.strip() for l in languages.split(',')]
    return [(l, SUPPORTED_LANGUAGES[l]) for l in languages if l in SUPPORTED_LANGUAGES]


def _get_default_password_min_length(fallback):
    return 10 if environ.get('USE_RECOVERY_CODE', str(bool(fallback))).lower() == 'true' else 6


class Configuration(metaclass=MetaFlaskEnv):
    VERSION = '0.9.5'

    PORT = 8000
    SECRET_KEY = 'dummy-secret'
    SITE_TITLE = 'Login Test Site'
    LANGUAGES = 'en'  # separated by a comma, for example 'en,bg'
    USE_RECOVERY_CODE = True
    SUBJECT_PREFIX = ''
    ABOUT_URL = 'https://github.com/epandurski/hydra_login2f'
    STYLE_URL = ''
    LOGIN_PATH = '/login'
    CONSENT_PATH = '/consent'
    HYDRA_ADMIN_URL = 'http://hydra:4445'
    REDIS_URL = 'redis://localhost:6379/0'
    SQLALCHEMY_DATABASE_URI = ''
    SQLALCHEMY_POOL_SIZE = None
    SQLALCHEMY_POOL_TIMEOUT = None
    SQLALCHEMY_POOL_RECYCLE = None
    SQLALCHEMY_MAX_OVERFLOW = None
    MAIL_SERVER = 'localhost'
    MAIL_PORT = 25
    MAIL_USE_TLS = False
    MAIL_USE_SSL = False
    MAIL_USERNAME = None
    MAIL_PASSWORD = None
    MAIL_DEFAULT_SENDER = None
    MAIL_MAX_EMAILS = None
    MAIL_ASCII_ATTACHMENTS = False
    RECAPTCHA_PUBLIC_KEY = '6Lc902MUAAAAAJL22lcbpY3fvg3j4LSERDDQYe37'
    RECAPTCHA_PIVATE_KEY = '6Lc902MUAAAAAN--r4vUr8Vr7MU1PF16D9k2Ds9Q'
    RECAPTCHA_REQUEST_TIMEOUT_SECONDS = 5

    RECAPTCHA_CHALLENGE_URL = 'https://www.google.com/recaptcha/api.js'
    RECAPTCHA_VERIFY_URL = 'https://www.google.com/recaptcha/api/siteverify'
    HYDRA_REQUEST_TIMEOUT_SECONDS = 5
    SEND_USER_UPDATE_SIGNAL = False
    SHOW_CAPTCHA_ON_SIGNUP = True
    CAPTCHA_RESPONSE_FIELD_NAME = 'g-recaptcha-response'
    LOGIN_VERIFIED_DEVICES_MAX_COUNT = 10
    LOGIN_VERIFICATION_CODE_EXPIRATION_SECONDS = 60 * 60
    MAX_LOGINS_PER_MONTH = 10000
    SECRET_CODE_MAX_ATTEMPTS = 10
    SIGNUP_REQUEST_EXPIRATION_SECONDS = 24 * 60 * 60
    CHANGE_EMAIL_REQUEST_EXPIRATION_SECONDS = 24 * 60 * 60
    CHANGE_RECOVERY_CODE_REQUEST_EXPIRATION_SECONDS = 60 * 60
    LANGUAGE_COOKE_NAME = 'user_lang'
    COMPUTER_CODE_COOKE_NAME = 'user_cc'
    LOGIN_VERIFICATION_COOKE_NAME = 'user_lv'
    PASSWORD_HASHING_METHOD = '6'
    PASSWORD_MIN_LENGTH = _get_default_password_min_length(USE_RECOVERY_CODE)
    PASSWORD_MAX_LENGTH = 64
    SEND_FILE_MAX_AGE_DEFAULT = 12096000
    MAX_CONTENT_LENGTH = 1024
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    SQLALCHEMY_ECHO = False
    BABEL_DEFAULT_LOCALE = 'en'
    BABEL_DEFAULT_TIMEZONE = 'UTC'
    LANGUAGE_CHOICES = _get_language_choices(LANGUAGES)
